Source/WebCore/html/HTMLMediaElement.cpp

```diff
@@ -1678,55 +1681,105 @@ void HTMLMediaElement::loadResource(const URL& initialURL, const ContentType& co
+        }
 #endif
 
-    if (!loadAttempted && m_blob && !m_remotePlaybackConfiguration) {
-        loadAttempted = true;
-        ALWAYS_LOG(LOGIDENTIFIER, "loading generic blob");
-        if (!m_blobURLForReading.isEmpty())
-            ThreadableBlobRegistry::unregisterBlobURL(m_blobURLForReading);
-        m_blobURLForReading = { BlobURL::createPublicURL(document().protectedSecurityOrigin().ptr()), document().topOrigin().data() };
-        ThreadableBlobRegistry::registerBlobURL(document().protectedSecurityOrigin().ptr(), document().policyContainer(), m_blobURLForReading, m_blob->url());
+        return player->load(url, contentType, keySystem, !!m_remotePlaybackConfiguration);
+    };
 
-        if (!player->load(m_blobURLForReading, contentType.isEmpty() ? ContentType { m_blob->type() } : contentType, keySystem, !!m_remotePlaybackConfiguration))
-            mediaLoadingFailed(MediaPlayer::NetworkState::FormatError);
-    }
+    if (contentType.isEmpty() && url.protocolIsData())
+        contentType = ContentType(mimeTypeFromDataURL(url.string()));
 
-    if (!loadAttempted && !player->load(url, url.protocolIsBlob() && contentType.isEmpty() ? ContentType { ThreadableBlobRegistry::blobType(url) } : contentType, keySystem, !!m_remotePlaybackConfiguration))
+    bool needContentType =
+#if ENABLE(MEDIA_SOURCE)
+        !m_mediaSource &&
+#endif
+#if ENABLE(MEDIA_STREAM)
+        !m_mediaStreamSrcObject &&
+#endif
+        !m_remotePlaybackConfiguration;
+
+    if (needContentType && contentType.isEmpty()) {
+        // Attempt to determine the content-type from the URL file extension.
+        contentType = ContentType { emptyString(), url };
+        if (contentType.isEmpty()) {
+            sniffForContentType(url, WTFMove(completionHandler));
+            return;
+        }
+    }
+    if (!completionHandler(contentType))
         mediaLoadingFailed(MediaPlayer::NetworkState::FormatError);
+    else
+        mediaPlayerRenderingModeChanged();
+}
 
-    mediaPlayerRenderingModeChanged();
+void HTMLMediaElement::sniffForContentType(const URL& url, Function<bool(const ContentType&)>&& completionHandler)
+{
+    auto handler = [this, completionHandler = WTFMove(completionHandler)](const ContentType& contentType) {
+        if (!completionHandler(contentType))
+            mediaLoadingFailed(MediaPlayer::NetworkState::FormatError);
+        else
+            mediaPlayerRenderingModeChanged();
+    };
+    m_sniffingAttempted = true;
+    ResourceRequest request(url);
+    request.setAllowCookies(true);
+    // https://mimesniff.spec.whatwg.org/#reading-the-resource-header defines a maximum size of 1445 bytes fetch.
+    m_sniffer = MediaResourceSniffer::create(*mediaPlayerCreateResourceLoader(), WTFMove(request), 1445);
+    if (!m_sniffer)
+        return handler(ContentType { });
+
+    m_sniffer->promise()->whenSettled(RunLoop::main(), [weakThis = WeakPtr { *this }, completionHandler = WTFMove(handler)](auto&& result) mutable {
```

Comment:
It doesn't seem like this lambda needs to capture `weakThis` at all. I think it is the `completionHandler` that should capture `weakThis` (and it actually currently captures `this` and uses it) and do the null check.

Otherwise, the completion handler capturing `this` in the function above looks unsafe and is only not crashing because of a weakPtr check at the call site, which is too fragile.

Reply:
That was the purpose of taking the weakPtr ; `sniffForContentType` is the one ensuring that the completionHandler is only called if `this` is valid and that the operation wasn't cancelled.

It may "look" unsafe ; but it really isn't

Reply:
I could change the name of `sniffForContentType` to indicate; like what we've done in other places like `ensureWeakOnHTMLMediaElement`

Reply:
As I commented elsewhere, I think `completionHandler` should always be called so 🤷🏻

Reply:
ok, I'll stop using a completionHandler then and use a full promise path, so that it can be rejected.

Reply:
CompletionHandler could be called with a std::expected or std::optional or ExceptionOr<> too.

Reply:
done